### PR TITLE
Fix race condition in custom libbeat instrumentation

### LINKED ISSUE #8383: "Output Events Rate" in stack monitoring is always zero
**APM Server version** (`apm-server version`): 8.3.0-BC4

**Description of the problem including expected versus actual behavior**:

"Output Events Rate" in stack monitoring is always zero.

**Steps to reproduce**:

 1. Start 8.3.0-BC4 with stack monitoring enabled.
 2. Send some events, check that they show up in the APM UI.
 3. Navigate to stack monitoring, observe the "Output Events Rate" chart is always reporting zero.

image

### PR DESCRIPTION
**Pull request by lahsivjar**
## Motivation/summary

The PR fixes race condition in libbeat instrumentation that surfaces during reload (initial and on subsequent configuration change). The race condition is in setting up custom libbeat instrumentation by using beats `monitoring#DefaultRegistry`.  The race arises from two reload events captured by APM-Server for every reload operation including the initial config load. The two reloads are used by APM-Server to load `inputs` using the beats reload registry and `output` using beats `OutputConfigReloader`.

Specifically, the race condition is in `serverRunner#newFinalBatchProcessor` (called from `serverRunner#run`) due to updates to a global `monitoring#Default` registry and the two reloads explained above. Even though the access to the `monitoring#Default` is protected by mutex in `serverRunner#newFinalBatchProcessor`; it is still possible that `serverRunner#newFinalBatchProcessor` call for a temporary reload happens later for a reload event that happens earlier. This will cause `monitoring#Default` to be in an incorrect state:
1. For initial load, the incorrect state would be no custom instrumentation
2. For subsequent reloads, the incorrect state would be instrumentation with an incorrect model indexer

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
~~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~~
~~- [ ] Documentation has been updated~~

## How to test these changes

1. Run `TestFleetIntegrationMonitoring` multiple times and observe they are not flaky
2. APM-Server metrics for elasticsearch output events should always be present in stack monitoring

## Related issues
Closes #8383

### CODE REVIEW

File: internal/beater/beater.go [hunk 1]
```diff
@@ -312,14 +312,23 @@ func (r *reloader) reloadOutput(config *reload.ConfigWithMeta) error {
...
-	r.mu.Unlock()
 	return r.reload()
 }
 
-func (r *reloader) reload() error {
+func (r *reloader) reloadOnce() error {
 	r.mu.Lock()
 	defer r.mu.Unlock()
+
+	return r.reload()
```
Comment (axw) [inline internal/beater/beater.go:324]: Seeing as this is only called in one place, should we inline it there?

**Review by axw — APPROVED** [review]
Nice one! Thanks for tracking this down.

Maybe in the future we can look at splitting `serverRunner.run` into two methods: one which initialises things, called with a mutex; and one which actively runs, which may be concurrent.

File: internal/beater/beater.go (hunk 1 above)
Comment (marclop) [reply internal/beater/beater.go:324]: I think it's also called in `reloadOutput`, on line 317, right?

File: internal/beater/beater.go [hunk 2]
```diff
@@ -345,6 +354,17 @@ func (r *reloader) reload() error {
...
 	}()
+
+	// Wait for the new runner to start; this avoids the race condition in updating
+	// the monitoring#Deafult global registry inside the runner due to two reloads,
+	// one for the inputs and the other for the elasticsearch output
+	select {
+	case <-runner.done:
+		return errors.New("runner exited unexpectedly")
+	case <-runner.started:
+		// runner has started
```
Comment (marclop) [inline internal/beater/beater.go:365]: Nice catch and great solution!

**Review by marclop — APPROVED** [review]
LGTM! Should this PR also be back ported to `8.4`? Seems like a good candidate to include with the TBS reload fix.

File: internal/beater/beater.go (hunk 1 above)
Comment (axw) [reply internal/beater/beater.go:324]: `reload` is called directly there, because `reloadOutput` already holds the mutex

Comment (lahsivjar) [reply internal/beater/beater.go]: Removed `reloadOnce` and inlined the mutex acquisition for standalone mode in `run`.

**Pull request merged**